externals/llvm-external-projects/torch-mlir-dialects/lib/Conversion/TcpToTosa/TcpToTosa.cpp

```diff
@@ -0,0 +1,113 @@
+#include "torch-mlir-dialects/Conversion/TcpToTosa/TcpToTosa.h"
+
+#include "torch-mlir-dialects/Dialect/Tcp/IR/TcpDialect.h"
+#include "torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.h"
+
+#include "../PassDetail.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+
+namespace mlir {
+
+#define GEN_PASS_DEF_CONVERTTCPTOTOSA
+#include "torch-mlir-dialects/Conversion/Passes.h.inc"
+
+namespace torch {
+namespace tcp {
+
+namespace {
+
+template<typename TcpOpT, typename TosaOpT>
+class ConvertElementwiseTcpOpToTosaOp : public OpConversionPattern<TcpOpT> {
+public:
+  using OpConversionPattern<TcpOpT>::OpConversionPattern;
+  using OpAdaptor = typename TcpOpT::Adaptor;
+  LogicalResult matchAndRewrite(
+            TcpOpT op,
+            OpAdaptor adaptor,
+            ConversionPatternRewriter& rewriter) const override {
+    llvm::SmallVector<mlir::Type> newTypes;
+    for (auto res : op->getResultTypes()) {
+      newTypes.push_back(OpConversionPattern<TcpOpT>::getTypeConverter()->convertType(res));
+    }
+    rewriter.replaceOpWithNewOp<TosaOpT>(
+          op, mlir::TypeRange(llvm::ArrayRef<mlir::Type>(newTypes)), adaptor.getOperands());
+    return success();
+  }
+};
+
+template<typename TcpOpT, typename TosaOpT>
+class ConvertMatmulTcpOpToTosaOp : public OpConversionPattern<TcpOpT> {
+public:
+  using OpConversionPattern<TcpOpT>::OpConversionPattern;
+  using OpAdaptor = typename TcpOpT::Adaptor;
+  LogicalResult matchAndRewrite(
+            TcpOpT op,
+            OpAdaptor adaptor,
+            ConversionPatternRewriter& rewriter) const override {
+    auto lhsTy = op.getLhs().getType().template cast<RankedTensorType>();
+    auto rhsTy = op.getRhs().getType().template cast<RankedTensorType>();
+    // TODO: Support 2D inputs.
+    if (lhsTy.getRank() != 3 || rhsTy.getRank() != 3) {
+      return op.emitError("tcp.matmul with inputs of rank != 3 cannot be legalized");
+    }
+    llvm::SmallVector<mlir::Type> newTypes;
+    for (auto res : op->getResultTypes()) {
+      newTypes.push_back(OpConversionPattern<TcpOpT>::getTypeConverter()->convertType(res));
+    }
+    rewriter.replaceOpWithNewOp<TosaOpT>(
+          op, mlir::TypeRange(llvm::ArrayRef<mlir::Type>(newTypes)), adaptor.getOperands());
+    return success();
+  }
+};
+
+class ConvertTcpToTosa : public ConvertTcpToTosaBase<ConvertTcpToTosa> {
+
+public:
+  void runOnOperation() override {
+    MLIRContext* context = &getContext();
+    RewritePatternSet patterns(context);
+    ConversionTarget target(*context);
+    target.addIllegalDialect<TcpDialect>();
```

Comment:
This will eventually have to go right since TCP won't fully convert to TOSA?  If yes, I suggest adding a comment here stating that.

Reply:
Yes, thats correct. Actually, I don't see a need for adding this, when we add the individual ops as legal and illegal below. So, removing this.